qualtran/serialization/data_types.py

```diff
@@ -86,3 +88,83 @@ def data_type_from_proto(serialized: SerializedDataTypes) -> QDType:
             "QAny, QInt, QIntOnesComp, QUInt, BoundedQUInt, "
             "QFixedPoint"
         )
+
+
+def int_or_sympy_from_qdt_proto(val: data_types_pb2.QDataType) -> Union[int, sympy.Expr]:
+    if val.HasField('qbit'):
+        return 1
+    if val.HasField('qany'):
+        return int_or_sympy_from_proto(val.qany.bitsize)
+    if val.HasField('qint'):
+        return int_or_sympy_from_proto(val.qint.bitsize)
+    if val.HasField('qintoc'):
+        return int_or_sympy_from_proto(val.qintoc.bitsize)
+    if val.HasField('quint'):
+        return int_or_sympy_from_proto(val.quint.bitsize)
+    if val.HasField('bquint'):
+        return int_or_sympy_from_proto(val.bquint.bitsize)
+    if val.HasField('qfxp'):
+        return int_or_sympy_from_proto(val.qfxp.bitsize)
+    raise ValueError(f"Cannot deserialize {val=}")
+
+
+def data_type_to_qdt_proto(data: QDType) -> data_types_pb2.QDataType:
```

Comment:
I think we can simply update  `data_type_to_proto`  and `data_type_from_proto` methods to expect and return a `QDType` instead of `SerializedDataTypes` and not have the two new methods implemented ?

Reply:
I guess my only question was if we ever would want the actual specific _pb2 data type (QAny, ...) vs the QDataType "wrapper".

Reply:
Given that they're not used outside of the tests my guess is no, but maybe?

Reply:
I think we'll always be dealing with `QDataType` wrapper in the proto serialization, so I vote for changing the code to deal with `QDataType` directly. 

We can consider making the existing functions internal helpers and changing your code to delegate to those helpers instead of rewriting the entire implementation. For example:

```python
def data_type_to_qdt_proto(data: QDType) -> data_types_pb2.QDataType:
    serialized_proto = data_type_to_proto(data)
    kwargs = {data_type_to_field_name[type(data)]:serialized_proto}
    return data_types_pb2.QDataType(**kwargs)

def data_type_from_qdt_proto(serialized: data_types_pb2.QDataType:) -> QDType:
    return data_type_from_proto(getattr(serialized, serialized.WhichOneof('val')))
```

Reply:
I went with the delete code suggestion!